Use equivalent PoW for non-main-chain requests

sipa opened the pull request:
In addition to checking whether the age of the requested block isn't too old, also check whether its cummulative proof of work is not more than 1 month behind on the best chain (at its current speed).

This should prevent a fingerprinting attack described by @sergiodemianlerner on #5820.

sipa committed:
Use equivalent PoW for non-main-chain requests

gmaxwell commented:
I'm not super enthusiastic of using work as a float here, but it seems correct and harmless.

utACK. (I don't currently have something setup to test this specific behavior.)

sipa commented:
Updated to use uint256 arithmetic.

laanwj commented:
utACK

sipa commented:
This patch is running on bitcoin.sipa.be, if someone wants to test whether syncing works.

gmaxwell commented:
ACK

sipa commented:
@sergiodemianlerner Feel like commenting?

SergioDemianLerner commented:
When *mi->second has greater nChainWork than *pindexBestHeader, then subtraction in GetBlockProofEquivalentTime() will overflow the uint256 and zero wrap. Multiplication by TargetSpacing() can also overflow the uint256. This will prevent the block from being served.
Of course, mi->second should never have more chain work than the tip (if it does it would become the new tip).
Nevertheless, as a protective layer against other bugs, I would add a condition that if mi->second has greater nChainWork than the tip, the block is always served.

gmaxwell commented:
@SergioDemianLerner how about an assert?  (if that case ever becomes possible, then something is horribly wrong; I think.)

SergioDemianLerner commented:
It seems that it is possible to have a block in mapBlockIndex with higher nChainWork than the best chain. This could happen if the node only received the header, but not the full block. Does this case apply here?

gmaxwell commented:
@SergioDemianLerner how about an assert?  (if that case ever becomes possible, then something is horribly wrong; I think.)

sdaftuar commented:
@gmaxwell  I'm not sure an assert makes sense here; the way the code seems to be structured now is that we don't generally hold a lock on cs_main in between updating mapBlockIndex (after receiving a block) and calling ActivateBestChain, so accessing chainActive.Tip() in between could result in an assertion failure.

I can't think of how this could occur outside of an interaction with the invalidateblock rpc call, but for example if you look at the implementation for invalidateblock, it calls InvalidateBlock while holding cs_main, then releases the lock and calls ActivateBestChain (which will lock cs_main itself where it needs to).  In between, it could be the case that some other VALID_SCRIPTS block has more work than chainActive.Tip().

Perhaps we should clarify precisely what the semantics ought to be for chainActive.Tip(): should it only be accessible if it reflects all known information about what the best chain should be?

sipa commented:
Let's just fix the code to deal with negatives.

sipa commented:
Updated.

morcos commented on src/main.cpp:
```diff
@@ -3487,11 +3486,13 @@ void static ProcessGetData(CNode* pfrom)
                         send = mi->second->IsValid(BLOCK_VALID_SCRIPTS) && (pindexBestHeader != NULL) &&
-                            (mi->second->GetBlockTime() > pindexBestHeader->GetBlockTime() - 30 * 24 * 60 * 60);
+                            (mi->second->GetBlockTime() > pindexBestHeader->GetBlockTime() - nOneMonth) &&
+                            (GetBlockProofEquivalentTime(*mi->second, *pindexBestHeader, *pindexBestHeader) > nOneMonth);
```
Is this supposed to be < nOneMonth?

sipa replied:
Ugh, yes. This is why code like this needs tests...

sipa commented:
* Rebased
* Changed the argument order of GetBlockProofEquivalentTime(a, b, c, d) to be identical to a->GetBlockTime () - b->GetBlockTime().
* Changed the order of calls in the fingerprinting test to be consistent with this.
* Passed Consensus::Params as an argument (make @jtimon happy).
* Added a simple unit test.

gavinandresen commented:
ACK

rebroad commented on src/pow.cpp:
```diff
@@ -114,3 +114,20 @@ arith_uint256 GetBlockProof(const CBlockIndex& block)
+        r = from.nChainWork - to.nChainWork;
+        sign = -1;
+    }
+    r = r * arith_uint256(params.nPowTargetSpacing) / GetBlockProof(tip);
```
nPowTargetSpacing, which always equals 600 regardless.... why can't this just be a global variable? It would save having to pass around the params everywhere.
